docs/theory.rst

```diff
@@ -0,0 +1,248 @@
+
+        def banner(s):
+            if trace:
+                print(f"\n{g._caller_name(2)}: ===== {s}")
+
+
+        banner('after setUp')
+
+        # 1. setUp creates self.project.
+
+            # setUp instantiates self.project to a Project instance.
+            # self.project = testutils.sample_project()
+
+        # 2. get_string_scope sets self.scope to the scope of the test string.
+
+            # Sets self.scope to pyobjectsdef.PyModule(project.pycore, code, ...)
+            # (code is the test string, defined above.)
+
+        # ??? Instantiating the pyobjectsdef.PyModule does all the work ???
+
+            # PyDefinedObject.__init__ calls:
+
+                # self.concluded_attributes = self.get_module()._get_concluded_data()
+                # self.attributes = self.get_module()._get_concluded_data()
+
+            # But all attributes are empty for this test.
+
+        scope = libutils.get_string_scope(self.project, code)
+
+        banner('after get_string_scope\n')
+
+            # scope is a GlobalScope.  It might be any subclass of Scope.
+            # scope.pyobject is a pyobjectsdef.PyModule.
+
+        # if trace: g.trace('*** scope.pyobject', scope.pyobject)
+
+        # *** Calling scope["Sample"] (via _ScopeVisitor._ClassDef)
+        #     instantiates pyobjects.PyClass *and* pyobjectsdef.PyClass.
+        #     (Because pyobjectsDef.PyClass is a subclass of pyobjects.PyClass.)
+        # scope["Sample"] is a pynamesdef.DefinedName.
+
+        sample_class = scope["Sample"].get_object()
+
+        banner('after sample_class = scope["Sample"].get_object()\n')
+
+
+            # sample_class is a pyobjectsdef.PyClass ("::Sample" at ...)
+            # scope["Sample"] is a DefinedName.
+            # scope["Sample"].pyobject is a pyobjectsdef.PyClass.
+
+        a_var = scope["a_var"].get_object()
+
+            # a_var is a pyobjects.PyObject
+            # a_var.get_type() is a pyobjectsdef.PyClass ("::Sample" at ...)
+
+            # scope["a_var"] is an pynamesdef.AssignedName.
+            # scope["a_var"].pyobject is a pynames._Inferred.
+            # scope["a_var"].get_object() is a pyobjects.PyObject.
+
+        if trace:
+            print('')
+            print(f"sample_class: {sample_class}")
+
+        self.assertEqual(sample_class, a_var.get_type())
+
+setUp simulates Rope's startup logic
+++++++++++++++++++++++++++++++++++++
+
+
+`ObjectInferTest.setUp` executes a simplified version of Rope's startup logic:
+
+.. code-block:: python
+
+    def setUp(self):
+        super().setUp()
+        self.project = testutils.sample_project()
+
+sample_project() just creates a `Project` instance.
+
+Rope's actual startup code is more complex, but for now I think it's safe to ignore these details.
+
+This code falls into `pdb` only for our test:
+
+.. code-block:: python
+
+    def setUp(self):
+        super().setUp()
+        if self.id().endswith('test_simple_type_inferencing'):
+            g.pdb()
```

Comment:
Rather than embedding the test name like this, usually if I need to drop down to a debugger for a particular test, I'd just drop a hard breakpoint using either the standard `breakpoint()/PYTHONBREAKPOINT` call or `from IPython import embed; embed()` or `from pudb import set_trace; set_trace()`, and then use a test runner to spawn individual pytest, e.g.

    $ pytest path/to/file.py:TestSomething.test_some_test_name

or using partial search path:

    $ pytest -k some_test_name

Reply:
@lieryan Thanks for these tips. I use a similar workflow now.